torch/csrc/jit/runtime/script_profile.cpp

```diff
@@ -0,0 +1,124 @@
+#include <torch/csrc/jit/runtime/script_profile.h>
+
+#include <atomic>
+#include <chrono>
+#include <mutex>
+#include <unordered_set>
+
+#include <c10/util/Exception.h>
+#include <torch/csrc/jit/api/function_impl.h>
+
+namespace torch {
+namespace jit {
+
+namespace {
+
+class ProfilesRegistry {
+ public:
+  bool empty() {
+    return empty_.load(std::memory_order_relaxed);
+  }
+  void addProfile(ScriptProfile& p) {
+    std::lock_guard<std::mutex> g(mutex_);
+    enabledProfiles_.emplace(&p);
+    empty_.store(false, std::memory_order_relaxed);
+  }
+  void removeProfile(ScriptProfile& p) {
+    std::lock_guard<std::mutex> g(mutex_);
+    enabledProfiles_.erase(&p);
+    if (enabledProfiles_.empty()) {
+      empty_.store(true, std::memory_order_relaxed);
+    }
+  }
+  void send(profiling::Datapoint datapoint) {
+    std::lock_guard<std::mutex> g(mutex_);
+    for (auto* p : enabledProfiles_) {
+      p->addDatapoint(datapoint);
+    }
+  }
+
+ private:
+  std::atomic<bool> empty_{true};
+  std::mutex mutex_;
+  std::unordered_set<ScriptProfile*> enabledProfiles_;
+};
+
+ProfilesRegistry& getProfilesRegistry() {
+  static auto& registry = *new ProfilesRegistry{};
+  return registry;
+}
+
+} // namespace
+
+namespace profiling {
+
+/* static */ c10::optional<InstructionSpan> InstructionSpan::tryMake(
+    Node& node) {
+  if (getProfilesRegistry().empty()) {
+    return {};
+  }
+  return InstructionSpan{node};
+}
+
+InstructionSpan::InstructionSpan(Node& node) : datapoint_(node.sourceRange()) {}
+
+InstructionSpan::InstructionSpan(InstructionSpan&& other)
+    : empty_(other.empty_), datapoint_(std::move(other.datapoint_)) {
+  other.empty_ = true;
+}
+
+InstructionSpan::~InstructionSpan() {
+  if (empty_) {
+    return;
+  }
+  datapoint_.end = std::chrono::steady_clock::now();
+  getProfilesRegistry().send(std::move(datapoint_));
+}
+
+} // namespace profiling
+
+void ScriptProfile::enable() {
+  if (!std::exchange(enabled_, true)) {
+    getProfilesRegistry().addProfile(*this);
+  }
+}
+
+void ScriptProfile::disable() {
+  if (std::exchange(enabled_, false)) {
+    getProfilesRegistry().removeProfile(*this);
+  }
+}
+
+void ScriptProfile::addDatapoint(profiling::Datapoint datapoint) {
```

Comment:
I'm not sure whether passing by shared_ptr is more performant, because we have one more heap allocation in that case. Meanwhile we have num_active_profiles more copies for the current approach, but my assumption is small copies like this should be super fast in modern architectures, and 90% of the cases num_active_profile == 1, so in common cases 1 allocation should take longer than 1 copy.